python/ray/serve/_private/application_state.py

```diff
@@ -85,35 +109,71 @@ def deploy_obj_ref(self) -> Optional[ObjectRef]:
 
     @property
     def deployments(self) -> List[str]:
-        """Return all deployments name from the application"""
-        if self._deployment_params is None:
+        """Names of all deployments in application."""
+        if self._target_state.deployment_infos is None:
             return []
-        return [params["name"] for params in self._deployment_params]
+        return list(self._target_state.deployment_infos.keys())
+
+    def recover_target_state_from_checkpoint(self, checkpoint_data):
+        logger.info(f"Recovering target state for app {self._name} from checkpoint.")
+        self._target_state = checkpoint_data
+
+    def _set_target_state(
+        self,
+        deployment_infos: Optional[Dict[str, DeploymentInfo]] = None,
+        deleting: bool = False,
+    ):
+        if deleting:
+            target_state = ApplicationTargetState(dict(), True)
```

Comment:
what is the diff between dict() and None in `ApplicationTargetState` ?
If no, I feel we can set to `None` to avoid any confusion.

Reply:
`None` is for when a new config has been deployed and we're waiting on `deploy_obj_ref`. In this period we don't perform any reconciliation. I set target state to an empty dictionary so that in the reconciliation loop, all of the deployments are removed from the application. LMK if that's confusing.

Reply:
oh my question is do we need to have separate attribute value (None and dict()) to indicate different scenario inside the `ApplicationTargetState` ?

Reply:
The problem with only using one or the other is that when a new config is deployed, I want to stop reconciliation, which is different from setting to empty dictionary because that will remove all the deployments when a new config is deployed.